### ISSUE TITLE
Read direct from / write directly to TAP

### ISSUE THREAD

**Issue by rbradford** [cloud-hypervisor/cloud-hypervisor#1265]
Remove the intermediate buffer in the virtio-net implementation and instead read directly into the memory using `.read_from()`

Upon tap event: While queue has free descriptors and data in tap  read up to "descriptor length" from tap.  If queue runs out of free descriptors turn off tap listening. 

Upon RX queue event: Restore tap listening. And repeat as per tap event processing potentially turning off listening again.

TX queue event: Write descriptor directly to tap.

**Comment by rbradford** [comment]
@sboeuf @bonzini After much experimentation today I don't think it's possible to read directly from the tap device into the descriptor. That's because the the size of the buffer passed in must be at least sufficient carry the header and the frame itself otherwise TAP truncates the packet. However the header and the frame are in different descriptors in the descriptor chain.

This means if we try and read 12 bytes (size of header) from the tap into the first descriptor in the chain we lose the rest of that packet.

The sources I found for this are: https://stackoverflow.com/a/17140017:

> Reads and writes with TUN/TAP, much like reads and writes on datagram sockets, must be for complete packets. If you read into a buffer that is too small to fit a full packet, the buffer will be filled up and the rest of the packet will be discarded. For writes, if you write a partial packet, the driver will think it's a full packet and deliver the truncated packet through the tunnel device.

>Therefore, when you read a TUN/TAP device, you must supply a buffer that is at least as large as the configured MTU on the tun or tap interface.

This tun/tap Rust crate also suggests the same https://docs.rs/tun-tap/0.1.0/tun_tap/struct.Iface.html#method.recv

QEMU also uses an intermediate buffer:

```
/* Maximum GSO packet size (64k) plus plenty of room for
 * the ethernet and virtio_net headers
 */
#define NET_BUFSIZE (4096 + 65536)
```

@bonzini Do you have any other thoughts on how we can avoid a intermediate buffer?

**Comment by bonzini** [comment]
Yeah, looks like that is the case. Of course for the specific case of virtio-net one could just use vhost.
